캠핑용품 예약 및 결제 기능

Habist opened the pull request:
## 작업 내용
렌탈 등록된 캠핑용품을 예약하고 금액만큼 바로 결제를 진행하는 기능을 구현했습니다.

## 주의 사항
iamport 모듈로 결제 기능을 구현했습니다.

>  참조이슈
> #48

Habist committed:
feat: [#48] 렌탈 비용 검색 기능 추가

ID와 날짜를 기반으로 렌탈 비용을 검색할 수 있는 기능을 추가하였습니다.

Habist committed:
feat: [#48] 예약 테이블 스키마 추가

Habist committed:
feat: [#48] 예약 기능 추가

캠핑용품을 렌탈하기 위해 예약하는 기능을 추가하였습니다.

Habist committed:
feat: [#48] 예약 생성 기능 이용 시 로그인 검증 과정 추가

예약 생성 기능 이용 시 로그인이 필요하도록 애노테이션을 추가하였습니다.

Habist committed:
feat: [#48] 결제 검증 기능 추가

결제 후 결제 정보에 대한 검증을 진행할 수 있는 기능을 추가했습니다.

Habist committed:
chore: [#48] 결제 모듈 API KEY 속성 추가

Habist committed:
feat: [#48] 결제 페이지 폼 추가

결제를 직접 수행할 수 있는 페이지를 추가하였습니다.

Habist committed:
refactor: [#48] 프로퍼티 한글 제거 및 빈 테스트 제거

Habist commented:
지금 CI가 Pull Request Base 브랜치로만 진행되는 이슈가 있는데 이슈 새로 만들어서 문제 해결하도록 하겠습니다.
해결하는 동안은 PR CI 잠시 inactive 시켜도 괜찮을까요?

f-lab-dev commented:
네 그렇게 해주셔도 됩니다~

Habist commented:
Jenkins CI 다시 구성하였고 PR의 compare 브랜치로 빌드 잘 되네요~! 이슈 따로 만들지 않고 바로 해결했습니다!

Habist committed:
refactor: [#48] 중복 코드 제거

f-lab-dev commented on src/main/java/com/campool/model/CreateBookingRequest.java:
```diff
@@ -0,0 +1,26 @@
+    private final long rentalId;
+
+    @NonNull
+    @DateTimeFormat(pattern = "yyyy-MM-dd")
```
이거는 어노테이션을 달 필요 없이 default로 이렇게 받아올 수 있게 할 수 있는 방법이 없을까요?

Habist replied:
Formatter나 Converter를 이용한 방법을 말씀하시는 건가요?

f-lab-dev replied:
jackson의 `ObjectMapper`에 모듈을 추가할 수 있지 않을까요?

f-lab-dev commented on src/main/java/com/campool/service/BookingService.java:
```diff
@@ -0,0 +1,60 @@
+    public CreateBookingResponse createBooking(CreateBookingRequest request, String userId) {
+        int cost = getCostByIdAndDate(request.getRentalId(), request.getStartDate(), request.getEndDate());
+        int rentalPeriod = getCountByDate(request.getStartDate(), request.getEndDate());
+        int amount = cost * rentalPeriod;
```
이런 곳에 개행 하나 넣어주면 코드 가독성이 약간 향상될 것 같네요~

Habist replied:
리뷰 반영했습니다!

f-lab-dev commented on src/main/java/com/campool/service/PaymentService.java:
```diff
@@ -0,0 +1,70 @@
+
+    public Payment getPaymentByUid(String impUid) {
+        try {
+            return new IamportClient(apiKey, apiSecret).paymentByImpUid(impUid).getResponse();
```
1. 아임포트와 관련된 책임은 분리해서 캡슐화된 메소드로 제공하는게 좋지 않을까요?
2. 매번 새 클라이언트 객체를 생성하는데 저 클래스는 하나만 만들고 돌려쓸 수 있다면 돌려쓰는게 메모리를 아낄 수 있을 것 같습니다~

Habist replied:
> 1. 아임포트와 관련된 책임은 분리해서 캡슐화된 메소드로 제공하는게 좋지 않을까요?

먼저 요청 정보를 기반으로 결제 대행사로 결제정보를 호출하는 부분을 인터페이스에 의존하게 만들고 의존성 주입을 하도록 변경하였습니다. 

현재는 아임포트를 사용하고 있더라도 불가피한 상황에 다른 PG사로 변경이 필요하면 `PaymentService`가 `IamportClient`와 강하게 결합되어 있어 유연함이 떨어지네요.

Habist replied:
> 2. 매번 새 클라이언트 객체를 생성하는데 저 클래스는 하나만 만들고 돌려쓸 수 있다면 돌려쓰는게 메모리를 아낄 수 있을 것 같습니다~

클라이언트를 요청마다 생성하도록 한 이유는 `IamportClient` 내부 코드를 확인했을 때 OkHttpClient는 하나의 커넥션만 생성하고 통신 또한 동기적인 방식을 사용하여 요청/응답하도록 설계되어 있었습니다. 그래서 여러 요청이 동시에 들어오게 되면 동시성 성능이 매우 나빠질 것을 염려하여 계속 생성하도록 만들었습니다. 

물론 현재처럼 계속 생성하면 메모리 사용률이 올라가기 때문에 다른 방안을 찾아야 할 것 같은데
`IamportClient` 내부 소스를 변경하여 커넥션 풀로 만들도록 하거나 오브젝트 풀로 만들어 정적 팩토리 메소드로 제공하는건 어떨까요?
작업이 너무 번거롭거나 좋지 않은 방법같으면 그냥 말씀하신대로 성능을 포기하고 `IamportClient` 하나만 생성해서 사용하는 것이 좋을까요?

만약 제가 잘못 알고 있는 부분이 있으면 리뷰 부탁드립니다.

> [OkHttpClient 생성 코드](https://github.com/iamport/iamport-rest-client-java/blob/master/src/main/java/com/siot/IamportRestClient/IamportClient.java#L338)
> [execute(동기적인 통신) 사용 부분](https://github.com/iamport/iamport-rest-client-java/blob/master/src/main/java/com/siot/IamportRestClient/IamportClient.java#L96)

f-lab-dev replied:
아 그렇군요~ 그러면 어쩔 수 없긴한데 그때그때 커넥션을 생성하는건 비효율적으로 보여서, 방법을 찾아봐야겠네요
`ObjectPool`은 그렇게 좋은 방법이 아니긴 합니다 ㅎㅎ

Habist committed:
refactor: [#48 - 리뷰 반영] 가독성 향상을 위한 개행 추가

Habist committed:
refactor: [#48 - 리뷰 반영] Payment 요청 클래스 책임 분리

결제 검증 Clinet가 변경되더라도 유연하게 대응 가능하도록 변경하였습니다.
검증을 위해 Client로 결제정보를 가져오는 책임을 분리했습니다.

f-lab-dev reviewed (approved):
java time module 같은 것들은 별도 이슈로 빼서 작업해도 될 것 같네요~
